### ISSUE TITLE
Fix "Sort by: Published Date"

### ISSUE THREAD

**Issue by KatharineV** [notch8/adventist_knapsack#185]
# Summary

The "Sort by Published Date" (both ascending and descending) feature in the catalog search and in collections does not properly sort based on the "Publication date" metadata field. If this feature is sorting based on a different date field, perhaps that would explain the problem, but if so, then we still would request work on this ticket, as explained in the last paragraph. 

If the feature is meant to sort based on the "Publication date" metadata field, then we'd like the functionality to work when dates are entered in that field according to the system's requirements. In [this example collection](https://adl.s2.adventistdigitallibrary.org/collections/53dd6c1a-ebb4-420a-a286-29714d81bf9c?locale=en) on ADL staging, we used the UI to enter publication date metadata, so we assume the system entered the date in a format that it prefers to read. The back end gave us a calendar widget, which we used. The dates appear properly in each work's metadata. However, when we use the sort by published date feature in the collection, the works do not sort in date order. See screenshots below.

# Acceptance Criteria

- [x] All Sort By options should function correctly
  - [x] on the [catalog search page](https://demo.s2.adventistdigitallibrary.org/catalog?utf8=%E2%9C%93&locale=en&search_field=all_fields&q=)
  - [x] and from the [collection show page](https://demo.s2.adventistdigitallibrary.org/dashboard/collections/9e158232-fdc5-441c-a8e1-14087b9a710e?locale=en) (admin view)
  
### Catalog search

![Image](https://github.com/user-attachments/assets/6a2adcc9-9ca4-4bea-90f3-81b9e5bab349)

### Collection show pg (admin view)

![Image](https://github.com/user-attachments/assets/a59b3a3e-eaa1-499a-ad48-f5e49cda4e8d)

# Screenshots or Video

Proper date in work metadata: 
https://adl.s2.adventistdigitallibrary.org/concern/published_works/22255888_the_minnesota_worker_december_21_1898?locale=en

![Image](https://github.com/scientist-softserv/adventist-dl/assets/118196520/d0ac223b-14bf-4144-ba2b-2788900e6366)

Automatic relevant sorting in the collection:

![Image](https://github.com/scientist-softserv/adventist-dl/assets/118196520/e272f9db-5e76-4214-9c71-f172079ad024)

Sort By: Published Date (Ascending):

![Image](https://github.com/scientist-softserv/adventist-dl/assets/118196520/922d4a67-553b-4912-b170-4872d6b0daeb)

Sort By: Published Date (Descending):

![Image](https://github.com/scientist-softserv/adventist-dl/assets/118196520/f0d71f97-c2b2-43be-82ca-8c86fd3aa03d)

These examples clearly show that the dates (which also appear in the work titles) are not sorting in proper order. They don't even change between ascending and descending. The behavior is the same in the catalog search, which offers a publication date sort. 

If the feature is actually pulling from a different date field, we'd like to know which one is being referenced, what date format the system will recognize and use to sort properly, and we'd like the Sort By label to be renamed to match the metadata field being referenced.  

As a sidenote only tangentially related to this ticket (and perhaps headed for its own ticket someday, or could be rolled into general Hyku upgrades of the future), the Publication Date metadata field seems like a bad choice for the catalog and collection search sort by feature. Not all works even have this metadata field in their work type. For example, Images. The sort by date feature would be better if it pulled a metadata field that appears in all work types, and it would be nice if there was a note in the UI that mentions to users that this is the field for date sorting. With soooooo many date fields across work types, it's very confusing what is used for search facets and sort by features etc. Additionally, the UI could contain a note regarding what date formats the system will actually recognize and use to sort. Is this info already included in Hyku documentation? 
See this screenshot of the back end of the UI "add new work" for the kind of note I'm suggesting would include format info and info about sort and facet dependency.  Hope this makes sense.

![Image](https://github.com/scientist-softserv/adventist-dl/assets/118196520/868de6df-8dba-4b63-8729-beb43ef64c94)

# Testing Instructions

**Testing note:** the staging site was partially reindexed resulting in most existing records being updated but not all. Our assumption is that the cut over will handle the update of all works in production. Most records will pass QA.

Ensure that works & collections have "Created Date" filled in. If there is no Created Date, they will sort to the end of the list, whether sorting by ascending or descending. 

- Go to the catalog index page
- Use the drop down menu to sort by the following
  - [ ] Title
  - [ ] Author
  - [ ] Published date (asc)
  - [ ] Published date (desc)
- Go to the Dashboard > Collection > select a collection with associated works > scroll down to works index page
- Use the drop down menu to sort by the following
  - [ ] Title
  - [ ] Author
  - [ ] Published date (asc)
  - [ ] Published date (desc)

# Notes

This will require an update of data to remove ["0~"] from blank author (creator) fields.

**Comment by jillpe** [comment]
This was fixed in UTK

**Comment by sjproctor** [comment]
We identified the source of the sorting problem is coming from Hyku. We are making the changes there first.

https://github.com/samvera/hyku/pull/2316/files

**Comment by sjproctor** [comment]
- If a record does not have a author (creator) the field will have `["0~"]` due to an older version of the order_already gem.
- The newest version of the gem handles this issue. The gem version was updated in Hyku and in the submodule of this application. 
- However, existing records in ADL with `["0~"]` will not register the author (creator) field as blank and therefore do not display or sort properly.
- To fix this problem, all records with  `["0~"]` in the creator field need to be updated. This can be done with the following script:

```
concerns = Hyrax.config.curation_concerns.map{ |c| "#{c}Resource".safe_constantize } << CollectionResource

concerns.each do |concern|
  works = Hyrax.query_service.find_all_of_model(model: concern)
  works.each do |work|
    next unless work['creator'] == ["0~"]
    work.creator = []
    work.save
  end
end
```

**Comment by ShanaLMoore** [comment]
# QA Results: ⚠ PENDING

tested on STAGING

**Testing note:** the staging site was partially reindexed resulting in most existing records being updated but not all. Our assumption is that the cut over will handle the update of all works in production. Most records will pass QA.

- Go to the catalog index page
- Use the drop down menu to sort by the following
  - [ ] Title
  - [ ] Author
  - [ ] Published date (asc)
  - [ ] Published date (desc)
- Go to the Dashboard > Collection > select a collection with associated works > scroll down to works index page
- Use the drop down menu to sort by the following
  - [ ] Title
  - [ ] Author
  - [ ] Published date (asc)
  - [ ] Published date (desc)

**Comment by KatharineV** [comment]
Team, as you're fixing this ticket, can you also educate me on which date field is used for the Sort By Published Date feature? There are so many date fields and they vary between work types. I have never been clear on which field maps to this feature, so your insight will be valuable.

**Comment by ShanaLMoore** [comment]
cc @sjproctor could you provide some insight here? ^^

**Comment by sjproctor** [comment]
Yeah, there are definitely a lot of date fields! We are using the `created_ssi` for publication date.

**Comment by laritakr** [comment]
This shows as "Date Created" on the work's metadata

**Comment by KatharineV** [comment]
@laritakr Thank you! These are the instructions I needed. I've tried to find out before which field and which format the system prefers. The answer wasn't available when I asked the last couple times, so this is really great. I understand why the sort won't work unless we make revisions. We can default to mm/dd/yyyy, but (in case the community ever sits down to think about this) I would love to see modifications in future versions that allow for ISO standard dates, since mm/dd/yyyy is confusing across international boundaries. 

Well, forget that for now, and thank you again!

**Comment by laritakr** [comment]
> @laritakr Thank you! These are the instructions I needed. I've tried to find out before which field and which format the system prefers. The answer wasn't available when I asked the last couple times, so this is really great.

You're welcome! Glad I could add clarity. Based on what I have found, I am moving this ticket back to Client QA.

As an added note, I found that anything with no created date will sort at the end, whether using ascending OR descending. 

We could put in some date standardization as we index the date used for sorting. This would require a reindex of all works & collections. If this is a desired change, you would need to make a ticket for it.

**Comment by laritakr** [comment]
# QA Results: ✅  Passed

### COLLECTION PAGE ✅  

tested on [STAGING](https://demo.s2.adventistdigitallibrary.org/dashboard/collections/9e158232-fdc5-441c-a8e1-14087b9a710e?utf8=%E2%9C%93&sort=date_created_ssi+desc&per_page=10&locale=en)

- Go to the Dashboard > Collection > select a collection with associated works > scroll down to works index page
- Use the drop down menu to sort by the following
  - [x] Title 
  - [x] Author
  - [ ] Published date (asc) ❌ 
  - [ ] Published date (desc) ❌

**Comment by KatharineV** [comment]
Folks, does the date_created field also inform the date range facet on the catalog search page? 
![image](https://github.com/user-attachments/assets/3a7828d3-4539-4428-bf24-6bb1e078015c)

**Comment by laritakr** [comment]
> Folks, does the date_created field also inform the date range facet on the catalog search page?

Yes, date_created is also used to build the sorted date. It expects the date_created to be in the format yyyy-mm-dd to extract the year to be used for sorting.

**Comment by KatharineV** [comment]
I tested this ticket with a set of works (see [importer](https://adl.s2.adventistdigitallibrary.org/importers/357?locale=en)) deposited into [a single collection](https://adl.s2.adventistdigitallibrary.org/dashboard/collections/0cb127b8-2e38-4c22-8d4d-3e3d41e18107). I put date metadata into the date_created column of the CSV. The final version metadata CSV is attached to that importer. When I formatted dates as 5/1/2023, 5/2/2023, 5/3/2023 and so on, the sort by date feature did NOT work. However, when I entered just the year, the sort worked and put works in the expected order. So, the feature works as described earlier in this ticket. 

I tested the sort on the collection dashboard page, the public show page, and in the catalog search.

**Comment by KatharineV** [comment]
Tested in Spectrum.b2 and my date sorts worked.
